pkg/providers/vault/client.go

```diff
@@ -48,11 +65,148 @@ func NewSecretClient(config SecretConfig) (pkg.SecretClient, error) {
 		config.retryWaitPeriodTime = retryTimeDuration
 	}
 
-	return Client{
+	secretClient := Client{
 		HttpConfig: config,
 		HttpCaller: httpClient,
-	}, nil
+	}
+
+	if bkgCtx == nil {
+		bkgCtx = context.Background()
+	}
+	// if there is contex already associated with the given token,
+	// then we cancel it first
+	if cancel, exists := vaultTokenToCancelFunc[tokenStr]; exists {
+		cancel()
+	}
+
+	ctx, cancel := context.WithCancel(bkgCtx)
+	if err = refreshToken(secretClient, ctx, errChan); err != nil {
+		cancel()
+	} else {
+		vaultTokenToCancelFunc[tokenStr] = cancel
+	}
+
+	return secretClient, err
+}
 
+func refreshToken(secretClient Client, ctx context.Context, errChan chan<- error) error {
+	token := secretClient.HttpConfig.Authentication.AuthToken
+
+	tokenData, err := getTokenLookupResponseData(secretClient)
+
+	if err != nil {
+		return err
+	}
+
+	if !tokenData.Data.Renewable {
+		// token is not renewable, log warning and return
+		msg := fmt.Sprintf("warning- token '%s' is not renewable from the secret store", token)
+		log.Println(msg)
+		return nil
+	}
+
+	// the renew internval is half of period value
+	tokenPeriod := time.Duration(tokenData.Data.Period) * time.Millisecond
+	renewInterval := tokenPeriod / 2
+	if renewInterval <= 0 {
+		// no renew
+		msg := fmt.Sprintf("warning- token '%s' renewInternal is 0, no token renewal", token)
+		log.Println(msg)
+		return nil
+	}
+
+	ttl := time.Duration(tokenData.Data.Ttl) * time.Millisecond
+
+	// if the current time-to-live is already less than the half of period
+	// need to renew the token right away
+	if ttl <= renewInterval {
+		// call renew self api
+		log.Println("ttl already <= half of the renewal period")
```

Comment:
Should use channel here also so messages go logged via EdgeX. Maybe instead of err channel we use log channel which is passed level and message??

Reply:
Or perhaps use the general logging client that everyone else is using.

Reply:
Modules typically don't log message directly. If EdgeX community is ok with passing in a reference to the logger so you can log the message, then I good with that.

Reply:
Interesting.  I'd like to hear other opinions.  I would like to offer the opinion that the places where the module is logging in this PR are important events.

Reply:
Agreed. That is why we need some way to pass the message back to be logged by caller or allow module to pull in the logging client and log directly.

Reply:
Either way, what I heard is that we need some pass-in logger channel or logging client reference to do the logging in the NewSecretClient() constructor.

Reply:
Lenny, I think the error channel is still needed for the caller to take actions depending on cases. Passing in an extra logging client seems to close the gap on the logging side when it is needed from the background process side.  What do you all think?

Reply:
I good with that.